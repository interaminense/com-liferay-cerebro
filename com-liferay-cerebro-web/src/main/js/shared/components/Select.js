import Component, {Config} from 'metal-jsx';

class SelectItem extends Component {
	render() {
		return <option {...this.otherProps()}>{this.props.children}</option>;
	}
}

class Select extends Component {
	render() {
		const {children, showBlankOption} = this.props;

		return (
			<select class="form-control select-root" {...this.otherProps()}>
				{showBlankOption && <SelectItem />}

				{children}
			</select>
		);
	}
}

Select.PROPS = {
	showBlankOption: Config.bool().value(false)
};

Select.Item = SelectItem;

export default Select;