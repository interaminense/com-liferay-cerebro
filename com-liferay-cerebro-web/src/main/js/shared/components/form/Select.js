import autobind from 'autobind-decorator';
import Component, {Config} from './Component';
import getCN from 'classnames';
import {noop, uniqueId} from 'lodash';

import Label from './Label';
import Select from '../Select';
import validate, {checkRequired} from '../../util/form-validators';

class FormSelect extends Component {
	created() {
		this._id = uniqueId('FormSelect');
	}

	attached() {
		super.attached();

		this.props.children.forEach(child => {
			const {selected, value} = child.props;

			if (selected) {
				this.state.value = value;
			}
		});
	}

	@autobind
	handleChange(event) {
		this.state.value = event.target.value;

		if (this.state.show) {
			this.validateFormElement();
		}

		this.props.onChange(event.target.value);
	}

	isRequired() {
		const {validator} = this.props;

		return validator && checkRequired(validator);
	}

	validateFormElement(showErrors = true) {
		const {validator} = this.props;

		let retVal = true;

		if (validator) {
			const result = validate(this.state.value, validator).find(
				item => !item.valid
			);

			const valid = result ? result.valid : true;

			this.setState({
				message: result ? result.message : '',
				show: showErrors,
				valid
			});

			retVal = valid;
		}

		return retVal;
	}

	render() {
		const {
			props: {children, info, inline, label},
			state: {message, show, valid, value}
		} = this;

		const classes = getCN({
			'form-inline-group': inline,
			'has-danger': !valid,
			'has-success': valid && show
		});

		return (
			<div class={classes}>
				{label && (
					<Label
						for={this._id}
						info={info}
						required={this.isRequired()}
					>
						{label}
					</Label>
				)}

				<Select
					{...this.otherProps()}
					id={this._id}
					onChange={this.handleChange}
					value={value}
				>
					{children}
				</Select>

				{show && <div class="help-block">{message}</div>}
			</div>
		);
	}
}

FormSelect.PROPS = {
	info: Config.string(),
	initialValue: Config.string().value(''),
	inline: Config.bool().value(false),
	label: Config.string(),
	onChange: Config.func().value(noop),
	validator: Config.object()
};

FormSelect.STATE = {
	message: Config.value(''),
	show: Config.value(false),
	valid: Config.value(true),
	value: Config.value('')
};

FormSelect.Item = Select.Item;

export default FormSelect;