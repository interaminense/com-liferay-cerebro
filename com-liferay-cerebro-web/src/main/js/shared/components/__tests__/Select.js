import Select from '../Select';

describe('Select', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = new Select();

		expect(component).toMatchSnapshot();
	});

	it('should render with blank option', () => {
		component = new Select({showBlankOption: true});

		expect(component).toMatchSnapshot();
	});

	it('should render with children', () => {
		const children = [
			<Select.Item key={1}>{'1'}</Select.Item>,
			<Select.Item key={2}>{'2'}</Select.Item>,
			<Select.Item key={3}>{'3'}</Select.Item>
		];

		component = new Select({children});

		expect(component).toMatchSnapshot();
	});

	it('should render with child selected', () => {
		const children = [
			<Select.Item key={1}>{'1'}</Select.Item>,
			<Select.Item key={2}>{'2'}</Select.Item>,
			<Select.Item key={3} selected>
				{'3'}
			</Select.Item>
		];

		component = new Select({children});

		expect(component).toMatchSnapshot();

		expect(component.element.value).toBe('3');
	});
});