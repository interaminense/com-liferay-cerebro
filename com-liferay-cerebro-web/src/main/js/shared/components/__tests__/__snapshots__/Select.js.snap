// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render 1`] = `<select class="form-control select-root"></select>`;

exports[`Select should render with blank option 1`] = `
<select class="form-control select-root">
  <option></option>
</select>
`;

exports[`Select should render with child selected 1`] = `
<select class="form-control select-root">
  <option>1</option>
  <option>2</option>
  <option selected="">3</option>
</select>
`;

exports[`Select should render with children 1`] = `
<select class="form-control select-root">
  <option>1</option>
  <option>2</option>
  <option>3</option>
</select>
`;
